internal/engine/experiment/quicping/quicping.go

```diff
@@ -0,0 +1,479 @@
+			out <- requestInfo{raw: packet, t: sendTime, dstID: hex.EncodeToString(dstID), srcID: hex.EncodeToString(srcID), err: err}
+			i += 1
+		}
+	}
+}
+
+// receiver receives incoming server responses and
+// dissects the payload of the version negotiation response
+func (m *Measurer) receiver(
+	ctx context.Context,
+	pconn model.UDPLikeConn,
+	out chan<- responseInfo,
+	sess model.ExperimentSession,
+) {
+	for ctx.Err() == nil {
+		// read (timeout was set in Run)
+		buffer := make([]byte, 1024)
+		n, addr, err := pconn.ReadFrom(buffer)
+		respTime := time.Now()
+		if err != nil {
+			// stop if the connection is already closed
+			if errors.Is(err, net.ErrClosed) {
+				break
+			}
+			// store read failures and continue receiving
+			out <- responseInfo{t: respTime, err: err}
+			continue
+		}
+		resp := buffer[:n]
+
+		// dissect server response
+		supportedVersions, dst, err := m.DissectVersionNegotiation(resp)
+		if err != nil {
+			// the response was likely not the expected version negotiation response
+			sess.Logger().Infof(fmt.Sprintf("response dissection failed: %s", err))
+			out <- responseInfo{raw: resp, t: respTime, err: err}
+			continue
+		}
+		// propagate receive information
+		out <- responseInfo{raw: resp, t: respTime, dstID: hex.EncodeToString(dst), versions: supportedVersions}
+
+		sess.Logger().Infof("PING got response from %s", addr)
+	}
+}
+
+// Run implements ExperimentMeasurer.Run.
+func (m *Measurer) Run(
+	ctx context.Context,
+	sess model.ExperimentSession,
+	measurement *model.Measurement,
+	callbacks model.ExperimentCallbacks,
+) error {
+	host := string(measurement.Input)
+	// allow URL input
+	if u, err := url.ParseRequestURI(host); err == nil {
+		host = u.Host
+	}
+	service := net.JoinHostPort(host, m.config.port())
+	udpAddr, err := net.ResolveUDPAddr("udp4", service)
+	if err != nil {
+		return err
+	}
+	rep := m.config.repetitions()
+	tk := &TestKeys{
+		Domain:      host,
+		Repetitions: rep,
+	}
+	measurement.TestKeys = tk
+
+	// create UDP socket
+	pconn, err := m.config.networkLibrary().ListenUDP("udp", &net.UDPAddr{})
+	if err != nil {
+		return err
+	}
+	defer pconn.Close()
+
+	// set context and read timeouts
+	deadline := time.Duration(rep*2) * time.Second
+	pconn.SetDeadline(time.Now().Add(deadline))
+	ctx, cancel := context.WithTimeout(ctx, deadline)
+	defer cancel()
+
+	sendInfoChan := make(chan requestInfo)
+	recvInfoChan := make(chan responseInfo)
+	pingMap := make(map[string]*pingInfo)
+
+	// start sender and receiver goroutines
+	go m.sender(ctx, pconn, udpAddr, sendInfoChan, sess)
+	go m.receiver(ctx, pconn, recvInfoChan, sess)
+L:
+	for {
+		select {
+		case req := <-sendInfoChan: // a new ping was sent
+			if req.err != nil {
+				tk.Pings = append(tk.Pings, &SinglePing{
+					ConnIdDst:   req.dstID,
+					ConnIdSrc:   req.srcID,
+					Failure:     archival.NewFailure(req.err),
+					Request:     &model.ArchivalMaybeBinaryData{Value: string(req.raw)},
+					RequestTime: formatTime(req.t),
```

Comment:
I think we should represent the time like we do for other nettests, i.e., as seconds elapsed since the `MeasurementStartTimeSaved` field of the measurement. This applies to both requests and responses: we should be using `time.Since(t0).Seconds()` every where we want to store a time. (In the previous snippet `t0` is the moment where a measurement was started (i.e., `MeasurementStartTimeSaved`.)